elm/Swole/Components/MovementInput.elm

```diff
@@ -0,0 +1,112 @@
+module Swole.Components.MovementInput exposing
+    ( Msg
+    , Model
+    , view
+    , update
+    )
+
+import Html exposing (Html, program, div, input, text, button)
+import Html.Attributes exposing (placeholder, type_, value, id)
+import Html.Events exposing (onClick, onInput)
+import Task
+import Dom exposing (focus)
+import List.Extra as List
+
+import Helpers.List as List
+import Helpers.Events exposing (onDeleteWhen)
+import Swole.Types.Movement exposing (Movement)
+
+type alias Movements = List (Int, Movement)
+type alias Model = Movements
+
+type Msg
+    = NoOp
+    | MovementChanged (Int, Movement)
+    | AddMovement
+    | DeleteMovement Int
+
+view : Movements -> Html Msg
+view movements =
+    div []
+        (movementFields movements)
+
+update : Msg -> Movements -> (Movements, Cmd Msg)
+update msg movements =
+    case msg of
+        NoOp ->
+            (movements, Cmd.none)
+
+        MovementChanged (idx, m) ->
+            let
+                (newIdx, newMovements) = splitMovements (idx, m)
+                (before, after) = movementsAround idx movements
```

Comment:
I think we'd discussed this structure when we talked; would it work here? http://package.elm-lang.org/packages/rtfeldman/selectlist/latest

Reply:
Yeah I need to look into this. The docs... aren't great, so I'm not 100% sure how to _use_ this, but I'm very interested in removing (or extracting) all of my manual indexing code.

Reply:
Yeah, the concept is that it's a non-empty list with a focus, and the data structure is such that there's a list before the selected item (if it's empty, the first item in the "list" is selected), and there's a list of items after the selected item (if it's also empty, there's only a single member in the list, the selected item).

I'm not sure if the underlying data structure here *requires* one item - if it does, this should work, but it might be outside the scope of this? There are a few functions I ended up having to write myself to make the package more usable:

```elm
selectSlide : (a -> comparable) -> SelectList a -> a -> SelectList a -- this selects a new member, with the understanding there's an accessor function to identify `a`
replaceSelected : SelectList a -> a -> SelectList a -- this swaps out the current selected value
buildSelectList : a -> List a -> SelectList a
deleteSlide : (a -> comparable) -> SelectList a -> a -> SelectList a -- same thing as `selectSlide`, given an accessor function that finds a particular item from `a` and the value you want to delete (`a`), it deletes the item and selects the closest one available. If there's only one value, it doesn't modify
```